contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,299 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(min_specialization)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+pub mod marketplace {
+    use access_control::{
+        traits::AccessControlled, Role, Role::Initializer, ACCESS_CONTROL_PUBKEY,
+    };
+    use game_token::TRANSFER_FROM_SELECTOR as TRANSFER_FROM_GAME_TOKEN_SELECTOR;
+    use ink_env::{
+        call::{build_call, Call, ExecutionInput, Selector},
+        CallFlags,
+    };
+    use ink_lang::{codegen::EmitEvent, reflect::ContractEventBase};
+    use ink_prelude::{format, string::String};
+    use openbrush::contracts::psp22::PSP22Error;
+    use ticket_token::{
+        BALANCE_OF_SELECTOR as TICKET_BALANCE_SELECTOR,
+        TRANSFER_SELECTOR as TRANSFER_TICKET_SELECTOR,
+    };
+
+    use crate::marketplace::Error::MissingRole;
+
+    type Event = <Marketplace as ContractEventBase>::Type;
+
+    const DUMMY_DATA: &[u8] = &[0x0];
+
+    #[ink(storage)]
+    pub struct Marketplace {
+        total_proceeds: Balance,
+        tickets_sold: Balance,
```

Comment:
I know it's just an alias, but it seems weird to have Balance here, for sth that is clearly a counter - maybe use u128 explicitly

Reply:
On the one hand... sure. On the other the tickets are a PSP22, so a number of tickets should have type balance... I went back and forth on this, don't have an opinion, so if this doesn't convince you the other way, I'll change it.